source/agora/node/TransactionRelayer.d

```diff
@@ -0,0 +1,462 @@
+            this.timers ~= this.taskman.setTimer(config.node.relay_tx_interval, &this.relayTransactions, Periodic.Yes);
+        this.timers ~= this.taskman.setTimer(1.minutes, &this.cleanRelayQueue, Periodic.Yes);
+    }
+
+    /// Called during node shutdown.
+    public void shutdown () @safe nothrow
+    {
+        foreach (timer; this.timers)
+            timer.stop();
+    }
+
+    /***************************************************************************
+
+        Returns an array of transactions ordered by fee, that needs to be
+        relayed to known network clients. The returned array might just be a
+        subset of all the transactions in the relay queue.
+
+        Returns:
+            an array of transactions ordered by fee
+
+    ***************************************************************************/
+
+    private Transaction[] getRelayTransactions () @safe nothrow
+    {
+        immutable now = this.clock.networkTime();
+        Transaction[] txs_gathered;
+        reserve(txs_gathered, this.config.node.relay_tx_max_num);
+
+        while (!this.txs.empty() &&
+                ((this.config.node.relay_tx_max_num == 0) ||
+                 (txs_gathered.length < this.config.node.relay_tx_max_num)))
+        {
+            TxHolder tx_holder = this.txs.front();
+            if (tx_holder.expiry >= now)
+            {
+                Transaction tx = this.pool.getTransactionByHash(tx_holder.hash);
+                if (tx != Transaction.init)
+                    txs_gathered ~= tx;
+            }
+
+            this.tx_hashes.remove(tx_holder.hash);
+            this.txs.removeFront();
+        }
+        return txs_gathered;
+    }
+
+    /// Sends an array of transactions ordered by fee to known network clients.
+    public void relayTransactions () @safe nothrow
+    {
+        this.getRelayTransactions().each!(tx => (*this.clients).each!(node_info => node_info.client.sendTransaction(tx)));
+    }
+
+    /// Cleans expired entries from the internal datastructures.
+    private void cleanRelayQueue () @safe
+    {
+        immutable now = this.clock.networkTime();
+
+        // Cannot iterate through a RedBlackTree, and delete its elements at the
+        // same time, so this is done in 2 steps.
+        auto tx_holders = this.txs[].filter!(tx_holder => tx_holder.expiry < now).array();
+        tx_holders.each!((tx_holder)
+        {
+            this.tx_hashes.remove(tx_holder.hash);
+            this.txs.removeKey(tx_holder);
+        });
+    }
+
+    /***************************************************************************
+
+        Adds a transaction to the relay queue.
+
+        Params:
+            tx = transaction to be added to the relay queue
+
+        Returns:
+            null, if there was no error, otherwise string detailing the error
+
+    ***************************************************************************/
+
+    private string addTransactionImp (in Transaction tx) @safe nothrow
```

Comment:
There are two functions named `addTransactionImp` and `addTransaction` which is kind of confusing. Why don't you merge the two functions into one or name them differently?